### PR TITLE
Refactoring: loss functions

### PR DESCRIPTION
**Pull request by pzelasko**
This is a follow-up to #149, it brings us a step closer to encapsulating the loss/graph related code in well-defined abstractions. It shares some code and introduces `CTCLoss` and `LFMMILoss`. 

I don't know how to show you only the relevant set changes (it is based on the branch in #149, so it displays changes from it too) -- it will be possible after merging #149.

BTW I tested both PRs that they run for a few batches in LibriSpeech, there are no changes in logic, just the code structure, so it should be safe to merge... Don't merge this one before #149 though.

### CODE REVIEW

File: snowfall/objectives/common.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+from typing import Dict, List, Tuple
+
+
+def encode_supervisions(supervisions: Dict[str, Tensor]) -> Tuple[Tensor, List[str]]:
+    """
+    Encodes Lhotse's ``batch["supervisions"]`` dict into a pair of torch Tensor,
+    and a list of transcription strings.
+
+    The supervision tensor has shape ``(batch_size, 3)``.
+    It's second dimension contains information about sequence index [0],
```
Comment (csukuangfj) [inline snowfall/objectives/common.py:13]: typo: It's -> Its

File: snowfall/objectives/ctc.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        super().__init__()
+        self.graph_compiler = graph_compiler
+
+    def forward(
+            self,
+            nnet_output: torch.Tensor,
+            texts: List,
+            supervision_segments: torch.Tensor
+    ) -> Tuple[torch.Tensor, int, int]:
+        num = self.graph_compiler.compile(texts)
```
Comment (csukuangfj) [inline snowfall/objectives/ctc.py:31]: Can we call the return value of `self.graph_compiler.compile(texts)` `num_graphs`
and the return value of `k2.intersect_dense(num, dense_fsa_vec, 10.0)` `num_lats`?

File: snowfall/training/mmi_graph.py [hunk 3]
```diff
@@ -59,23 +60,21 @@ def __init__(self,
...
 
         self.L_inv = L_inv
-        self.phones = phones
-        self.words = words
-        self.oov_id = self.words[oov]
+        self.oov_id = self.lexicon.words[oov]
         self.device = device
 
-        phone_symbols = get_phone_symbols(phones)
+        phone_symbols = get_phone_symbols(self.lexicon.phones)
```
Comment (csukuangfj) [inline snowfall/training/mmi_graph.py:77]: I think `self.lexicon.phone_symbols()` implements what `get_phone_symbols` is doing.

**Comment by csukuangfj** [comment]
Looks also great to me.

File: snowfall/training/mmi_graph.py (hunk 3 above)
Comment (pzelasko) [reply snowfall/training/mmi_graph.py:77]: Yeah I copied that function here - I’m trying to strike a balance between moving forward and not breaking every other training script that I didn’t update. I think after a bit more work we’ll be able to remove the left over code safely.

**Comment by danpovey** [comment]
This and the other PR look good to me (although I don't have time to go through them in much detail).
I don't think any of the reset of us have time to think too deeply about issues related to this, because we all
have short-term things we are working on.

What do you think about merging this to master, and also copying the current master to a branch called,
say, `pre_refactor`, and any work that we have that we've been working on in the context of the old master,
we can just commit to the `pre_refactor` branch and ask Piotr to help port to the newer code?

**Comment by pzelasko** [comment]
Sounds good to me, will do.

**Comment by danpovey** [comment]
Great, thanks!

**Comment by pzelasko** [comment]
The `pre_refactor` branch is available now, and I also created a git tag `v0.1-pre_refactor` (easy to find here: https://github.com/k2-fsa/snowfall/tags)

**Comment by danpovey** [comment]
great!  you can merge your changes.if.you havent already...

**Comment by pzelasko** [comment]
Alright, both MMI and CTC attention scripts for LibriSpeech are working. Merging now.
